File: Algorithms/Medium/3Sum.py
# Time complexity: O(n^2)
# Space complexity: O(1)
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        result = []
        nums.sort()

        n = len(nums)
        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            l, r = i + 1, n - 1
            while l < r:
                curr_sum = nums[i] + nums[l] + nums[r]
                if curr_sum < 0:
                    l += 1
                elif curr_sum > 0:
                    r -= 1
                else:
                    result.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1
                    l += 1
                    r -= 1
        
        return result

File: Algorithms/Medium/15.3Sum/3Sum.go
package medium

import "sort"

// Time complexity: O(n^2)
// Space caomplexity: O(1)
func threeSum(nums []int) [][]int {
	result := make([][]int, 0)
	sort.Ints(nums)

	for i := range nums {
		if nums[i] > 0 {
			return result
		}

		if i > 0 && nums[i-1] == nums[i] {
			continue
		}

		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if sum < 0 {
				l++
			} else if sum > 0 {
				r--
			} else {
				result = append(result, []int{nums[i], nums[l], nums[r]})
				for l < r && nums[l] == nums[l+1] {
					l++
				}

				for l < r && nums[r] == nums[r-1] {
					r--
				}

				l++
				r--
			}
		}
	}

	return result
}

// [-4,-1,-1,-1,0,1,2,2]

// [-1,0,1,2,-1,-4]

// [-4,-1,-1,0,1,2]
// 	 ^  ^        ^  -> -3

// [-4,-1,-1,0,1,2]
// 	 ^     ^     ^  -> -3

// [-4,-1,-1,0,1,2]
// 	 ^       ^   ^  -> -2

// [-4,-1,-1,0,1,2]
// 	 ^     	   ^ ^  -> -1

// [-4,-1,-1,0,1,2]
// 	    ^  ^     ^  -> 0 [-1,-1,2]

// [-4,-1,-1,0,1,2]
// 	    ^    ^   ^  -> 1

// [-4,-1,-1,0,1,2]
// 	    ^    ^ ^    -> 0 [-1,0,1]

// [-4,-1,-1,0,1,2]
// 	    ^    ^ ^    -> 0 [-1,0,1]
